### PR TITLE
[dagster-airflow] refactor airflow_db resources

### PR DESCRIPTION
**Pull request by Ramshackle-Jamathon**
### Summary & Motivation

This PR:
- adds a new api for creating an emphemeral airflowdb resource def `make_ephemeral_airflow_db_resource`
- drops setting any resource_defs from asset apis
- consolidates dag/dagrun resources into the op + airflowdb resource

### How I Tested These Changes
- existing unit coverage

### CODE REVIEW

File: python_modules/libraries/dagster-airflow/dagster_airflow/resources/airflow_ephemeral_db.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                importlib.reload(airflow.configuration)
+                importlib.reload(airflow.settings)
+                importlib.reload(airflow)
+            else:
+                importlib.reload(airflow)
+            if not airflow_initialized:
+                db.initdb()
+                create_airflow_connections(
+                    [Connection(**c) for c in context.resource_config["connections"]]
+                )
```
Comment (schrockn) [inline python_modules/libraries/dagster-airflow/dagster_airflow/resources/airflow_ephemeral_db.py:67]: recommend not doing anything this stateful and heavy in an `__init__` method. It makes classes much more difficult to test and reuse in other contexts. I would instead have a factory function/static method that takes the context, does the super heavy global state stuff, and gets only the data it needs out of the context.

File: python_modules/libraries/dagster-airflow/dagster_airflow/resources/airflow_ephemeral_db.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                        state=State.RUNNING,
+                        execution_date=execution_date,
+                    )
+            return dagrun
+
+
+def airflow_ephemeral_db(
+    context: InitResourceContext,
+) -> Generator[AirflowEphemeralDatabase, None, None]:
+    yield AirflowEphemeralDatabase(context)
```
Comment (schrockn) [inline python_modules/libraries/dagster-airflow/dagster_airflow/resources/airflow_ephemeral_db.py:146]: you can just return here

**Comment by schrockn** [comment]
Threw my suggested refactor in https://github.com/dagster-io/dagster/pull/12205

**Review by schrockn — APPROVED** [review]
Cool thanks for banging this out.

File: examples/with_airflow/with_airflow/assets_repository.py [hunk 3]
```diff
@@ -40,6 +39,8 @@ def new_upstream_asset_2():
 )
 
 
-@repository
+@repository(
+    top_level_resources={"airflow_db": make_ephemeral_airflow_db_resource()},
```
Comment (schrockn) [inline examples/with_airflow/with_airflow/assets_repository.py:43]: oh lord don't use this. @benpankow  we should make this `_top_level_resources` so indicate internal-only?

@Ramshackle-Jamathon move this to `Definitions` please

Comment (Ramshackle-Jamathon) [reply examples/with_airflow/with_airflow/assets_repository.py]: with_airflow is such a mess atm

**Review by schrockn — APPROVED** [review]
image

**Pull request merged**